guide/src/reference/attributes/on-js-imports/host_binding.md

```diff
@@ -0,0 +1,139 @@
+called is hoisted out of the generated shim and is bound to always be
+`Foo.prototype.bar`. This then uses the `Function.call` method to invoke that
+function with `getObject(arg0)` as the receiver.
+
+But wait, there's still a JS shim here even with `host_binding`! That's true,
+and this is simply a fact of future WebAssembly proposals not being implemented
+yet. The semantics, though, match the future [host bindings
+proposal][host-bindings] because the method being called is determined exactly
+once, and it's located on the prototype chain rather than being resolved at
+runtime when the function is called.
+
+## Interaction with future proposals
+
+If you're curious to see how our JS shim will be eliminated entirely, let's take
+a look at the generated bindings. We're starting off with this:
+
+```js
+const __wbg_bar_target = Foo.prototype.bar;
+
+export function __wbg_bar_a81456386e6b526f(arg0, arg1, arg2) {
+    let varg1 = getStringFromWasm(arg1, arg2);
+    return addHeapObject(__wbg_bar_target.call(getObject(arg0), varg1));
+}
+```
+
+... and once the [reference types proposal][reference-types] is implemented then
+we won't need some of these pesky functions. That'll transform our generated JS
+shim to look like:
+
+```js
+const __wbg_bar_target = Foo.prototype.bar;
+
+export function __wbg_bar_a81456386e6b526f(arg0, arg1, arg2) {
+    let varg1 = getStringFromWasm(arg1, arg2);
+    return __wbg_bar_target.call(arg0, varg1);
+}
+```
+
+Getting better! Next up we need the host bindings proposal. Note that the
+proposal is undergoing some changes right now so it's tough to link to reference
+documentation, but it suffices to say that it'll empower us with at least two
+different features.
+
+First, host bindings promises to provide the concept of "argument conversions".
+The `arg1` and `arg2` values here are actually a pointer and a length to a utf-8
+encoded string, and with host bindings we'll be able to annotate that this
+import should take those two arguments and convert them to a JS string (that is,
+the *host* should do this, the WebAssembly engine). Using that feature we can
+futher trim this down to:
+
+```js
+const __wbg_bar_target = Foo.prototype.bar;
+
+export function __wbg_bar_a81456386e6b526f(arg0, varg1) {
+    return __wbg_bar_target.call(arg0, varg1);
+}
+```
+
+And finally, the second promise of the host bindings proposal is that we can
+flag a function call to indicate the first argument is the `this` binding of the
+function call. Today the `this` value of all called imported functions is
+`undefined`, and this flag (configured with host bindings) will indicate the
+first argument here is actually the `this`.
+
+With that in mind we can further transform this to:
+
+```js
+export const __wbg_bar_a81456386e6b526f = Foo.prototype.bar;
+```
+
+and voila! We, with [reference types][reference-types] and [host
+bindings][host-bindings], now have no JS shim at all necessary to call the
```

Comment:
I would still call that a JS shim. It is JS glue code which is needed to convert a JS value (`Foo.prototype.bar`) into something which is capable of being imported by wasm, since wasm cannot import it directly.

Also, I think it should mention what the `structural` version would look like:

```js
export const __wbg_bar_a81456386e6b526f = function (varg1) { return this.bar(varg1); };
```

Quite short and simple, and also just as performant.

Reply:
I've switched to phrasing this as "JS function shim", I think being pedantic here about what's a JS shim and what isn't is detracting from the point of the documentatino. 

And again to be clear, this isn't trying to describe the performance today, it's describing the eventual performance where if you have *any JS shim at all* then wasm engines will be unable to route calls to C++ dom functions directly to C++. The whole point of remove the JS shim is to ensure that DOM calls can go straight to C++ with as little fuss as possible.

Reply:
Oh! I see! I finally understand our confusion.

You've been thinking about this as exclusively about JS built-ins (like `Node.prototype.appendChild`), whereas I've been thinking about it as a general purpose mechanism (which can be used on anything, including JS methods).

From the perspective of built-ins, you're completely right that adding in any sort of indirection (i.e. anything other than `export const bar = Foo.prototype.bar;`) is bad for performance, since it can't directly call the built-in C++ function (at least not without a bit of VM magic).

But on the other hand, from the perspective of regular JS classes, you won't gain any performance by doing `export const bar = Foo.prototype.bar;`, since it still has to call into a JS method.

So from a general standpoint, it is indeed a JS shim, but from the built-ins standpoint it isn't a JS shim. Hence our disagreement and confusion.

I think we could use a better term for it ("direct" vs "indirect" JS call, maybe? Or "built-in" vs "non-built-in" method?), but at least now the confusion has been resolved.

As for the name, none of them are great, and we can easily spend months bikeshedding it, so let's just go with `final`.

Reply:
Ah yes indeed, this would cause confusion! 

I don't know what this will mean for regular JS classes. I think it's too early to say whether it will or won't have a difference.